### ISSUE TITLE
Exhaustive switch returns remove narrowing

### ISSUE THREAD

**Issue by charles-toller** [microsoft/TypeScript#33733]
**TypeScript Version:** 3.7.0-dev.20191001 (also tested against typescript@beta)

**Search Terms:** exhaustive switch return undefined

**Code**

```ts
type Action = SetAction | ReplaceAction;
interface SetAction {
    type: 'ACTION_SET';
}
interface ReplaceAction {
    type: 'ACTION_REPLACE';
}
function test(state: {} | undefined, action: Action): {} {
    if (state === undefined) {
        return {};
    }
    switch (action.type) {
        case "ACTION_SET":
            return state;
        case "ACTION_REPLACE":
            return state;
    }
    return state;
}

```

(Run with `--strict`)

**Expected behavior:** No error, `state` should be narrowed to `{}` at the final return statement.

**Actual behavior:** test.ts:18:5 - error TS2322: Type '{} | undefined' is not assignable to type '{}'.
  Type 'undefined' is not assignable to type '{}'.

**Playground Link:** [link](https://www.typescriptlang.org/play/?noImplicitAny=false&strictFunctionTypes=false&strictPropertyInitialization=false&noImplicitThis=false&noImplicitReturns=false&target=99&jsx=2&ts=Nightly#code/C4TwDgpgBAggxsAlgewHZQLxQMoWPJNKAHygCUIwAbAQzggJVQG4AoRVYCAJwDM7oufAiZQA3qyhSooSAC4oAchgBhACoBJAPIA5APrYAomsVsAvu049+9cpVr1GRCdJngIC5eu36yhgAoAMqqGpqwWvACuqCJEXADOwAAUiTRcCmJmJFDRACYQvBwQuQA0UHSEqApOqACUGVku0oi8UCnAadAY3Tmo+YWoxbXikq7S3HiR3OiZbK4WrvEA7ojAcAAWbRVMAHSyEMNNY1BwNPHQAESqmroGxhdyo8euE8BT6Klcc8+n51BX3lufiCIQeT2eUle7ygnwg3ykC3Gk2mMI6X3CrCAA), but it currently isn't appearing on the nightly playgrounds (at least for me).

**Related Issues:** Bisecting reveals this started in 8cbf69489cd1974354ae07c81a4f03dc251d84b3, which was merged in #32695.

**Comment by jack-williams** [comment]
This probably falls under #26914.

**Comment by ahejlsberg** [comment]
This is an intended effect of #32695. It was always the case that we perform no control flow based narrowing in unreachable code (since it's not at all clear what assumptions to make about the flow of control in code that has no flow of control). For example, you'd see the same error if you placed a `throw` statement immediately before the final `return` statement. What's changed is that we now fully reflect the effects of exhaustive `switch` statements in control flow analysis--including the fact that they can cause code to become unreachable. We only got this partially right before.

If you really want the code to compile with the final `return` in place, you can convince the control flow analyzer that the code is reachable (even though it technically isn't) by including a `default` clause in the `switch` statement.

That said, I'd suggest compiling with `"allowUnreachableCode": "false"` so that you get errors on unreachable code (otherwise you only get "suggestions" which are reflected as dimmed out code in VS Code, but _aren't_ reported by the command-line compiler).

**Comment by charles-toller** [comment]
That makes sense. Thanks for the info!
